File: backend/app/models/news.py
from __future__ import annotations

import enum
from datetime import datetime
from typing import List, Optional

from sqlalchemy import String, Text, DateTime, Enum, ForeignKey, Boolean, Integer
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.database import Base


class Category(str, enum.Enum):
    POLITICS = "politics"
    ECONOMY = "economy"
    SOCIETY = "society"
    TECH = "tech"
    INTERNATIONAL = "international"
    UNCATEGORIZED = "uncategorized"


class SourceName(str, enum.Enum):
    GOOGLE_NEWS = "google_news"
    NAVER_NEWS = "naver_news"
    JOONGANG = "joongang"
    HANI = "hani"


class Source(Base):
    __tablename__ = "sources"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    name: Mapped[str] = mapped_column(Enum(SourceName), unique=True, nullable=False)
    display_name: Mapped[str] = mapped_column(String(100), nullable=False)
    base_url: Mapped[str] = mapped_column(String(255), nullable=False)
    is_active: Mapped[bool] = mapped_column(Boolean, default=True)
    last_crawled_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)

    articles: Mapped[List["Article"]] = relationship(back_populates="source")


class Article(Base):
    __tablename__ = "articles"

    id: Mapped[int] = mapped_column(Integer, primary_key=True)
    source_id: Mapped[int] = mapped_column(ForeignKey("sources.id"), nullable=False)
    title: Mapped[str] = mapped_column(String(500), nullable=False)
    url: Mapped[str] = mapped_column(String(1000), unique=True, nullable=False)
    summary: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    content: Mapped[Optional[str]] = mapped_column(Text, nullable=True)
    image_url: Mapped[Optional[str]] = mapped_column(String(1000), nullable=True)
    author: Mapped[Optional[str]] = mapped_column(String(200), nullable=True)
    category: Mapped[str] = mapped_column(
        Enum(Category), default=Category.UNCATEGORIZED, nullable=False
    )
    published_at: Mapped[Optional[datetime]] = mapped_column(DateTime, nullable=True)
    crawled_at: Mapped[datetime] = mapped_column(DateTime, default=datetime.utcnow)

    source: Mapped["Source"] = relationship(back_populates="articles")

File: frontend/src/types/news.ts
export type Category =
  | "politics"
  | "economy"
  | "society"
  | "tech"
  | "international"
  | "uncategorized";

export type SourceName =
  | "google_news"
  | "naver_news"
  | "chosun"
  | "joongang"
  | "hani"
  | "yonhap";

export interface Source {
  id: number;
  name: SourceName;
  display_name: string;
  base_url: string;
  is_active: boolean;
  last_crawled_at: string | null;
}

export interface Article {
  id: number;
  title: string;
  url: string;
  summary: string | null;
  image_url: string | null;
  author: string | null;
  category: Category;
  published_at: string | null;
  crawled_at: string;
  source: Source;
}

export interface ArticleList {
  total: number;
  page: number;
  page_size: number;
  items: Article[];
}

export interface ArticleFilters {
  page?: number;
  page_size?: number;
  category?: Category;
  source?: SourceName;
  q?: string;
}

export const CATEGORY_LABELS: Record<Category, string> = {
  politics: "정치",
  economy: "경제",
  society: "사회",
  tech: "기술",
  international: "국제",
  uncategorized: "기타",
};

export const SOURCE_LABELS: Record<SourceName, string> = {
  google_news: "Google 뉴스",
  naver_news: "네이버 뉴스",
  chosun: "조선일보",
  joongang: "중앙일보",
  hani: "한겨레",
  yonhap: "연합뉴스",
};
